tests/test_io.py

```diff
@@ -17,68 +25,146 @@ def case_trajectory_2D(self):
+            Nc=32, Ns=512, tilt="uniform", in_out=True
         ).astype(np.float32)
         return (
             "3D",
             trajectory,
             (0.23, 0.23, 0.1248),
-            (256, 256, 128),
-            True,
+            (64, 64, 32),
+            0,
             5,
-            10e3,
+            Gammas.Na,
             1.2,
         )
 
 
+@parametrize("gamma", [Gammas.Hydrogen])
+@parametrize("raster_time", [DEFAULT_RASTER_TIME])
+@parametrize_with_cases(
+    "kspace_loc, shape",
+    cases=[
+        CasesTrajectories.case_radial2D,
+        CasesTrajectories.case_radial3D,
+        CasesTrajectories.case_in_out_radial2D,
+    ],
+)
+@parametrize("gmax", [0.1, DEFAULT_GMAX])
+@parametrize("smax", [0.7, DEFAULT_SMAX])
+def test_trajectory_state_changer(kspace_loc, shape, gamma, raster_time, gmax, smax):
+    """Test the trajectory state changer."""
+    dimension = len(shape)
+    resolution = dimension * (0.23 / 256,)
+    trajectory = kspace_loc / resolution
+    gradients = np.diff(trajectory, axis=1) / gamma / raster_time
+    GS = get_gradient_amplitudes_to_travel_for_set_time(
+        kspace_end_loc=trajectory[:, 0],
+        end_gradients=gradients[:, 0],
+        gamma=gamma,
+        raster_time=raster_time,
+        gmax=gmax,
+        smax=smax,
+    )
+    # Hardware constraints check
+    assert np.all(np.abs(GS) <= gmax)
+    assert np.all(np.abs(np.diff(GS, axis=1) / raster_time) <= smax)
+    assert np.all(np.abs(GS[:, -1] - gradients[:, 0]) / raster_time < smax)
+    if np.all(trajectory[:, 0] == 0):
+        # If the trajectory starts at the origin, we can check that the first gradient is zero
+        assert np.all(GS.shape[1] < 10)
+    assert GS.shape[1] < 200  # Checks to ensure we don't have too many samples
+    # Check that ending location matches.
+    np.testing.assert_allclose(
+        np.sum(GS, axis=1) * gamma * raster_time,
+        trajectory[:, 0],
+        atol=1e-2 / min(resolution) / 2,
+    )
+    # Check that gradients match.
+    np.testing.assert_allclose(GS[:, 0], 0, atol=1e-5)
+
+    GE = get_gradient_amplitudes_to_travel_for_set_time(
+        kspace_start_loc=trajectory[:, -1],
+        kspace_end_loc=np.zeros_like(trajectory[:, -1]),
+        start_gradients=gradients[:, -1],
+        gamma=gamma,
+        raster_time=raster_time,
+        gmax=gmax,
+        smax=smax,
+    )
+    # Hardware constraints check
+    assert np.all(np.abs(GE) <= gmax)
+    assert np.all(np.abs(np.diff(GE, axis=1) / raster_time) <= smax)
+    assert np.all(np.abs(GE[:, -1]) / raster_time < smax)
+    assert GE.shape[1] < 200  # Checks to ensure we don't have too many samples
+    # Check that ending location matches.
+    np.testing.assert_allclose(
+        0,
+        trajectory[:, -1] + np.sum(GE, axis=1) * gamma * raster_time,
+        atol=1e-2 / min(resolution) / 2,
+    )
+    # Check that gradients match.
+    np.testing.assert_allclose(GE[:, 0], gradients[:, -1], atol=1e-5)
```

Comment:
Consider putting this in a dedicated test to have smaller tests ?

Reply:
I dont think i got you. Thing is the test takes time to run as it runs solver internally. I would prefer having all the checks in same test due to this. Lemme know if you know how to refactor to pre-calculate and then do checks in smaller tests

Reply:
you can use a fixture for the precomputation :)